gitpod-io/gitpod
# [supervisor] Support workspace ports

This PR moves the "served port" functionality from Theia to supervisor. Specifically it
- adds [new API to supervisor](https://github.com/gitpod-io/gitpod/blob/47f16bf2f9790dc4601b9e162ecd8b371961f3a8/components/supervisor-api/status.proto#L46-L54) which provides information about the ports currently serviced within the workspace
- provides Host-header rewriting proxying for services bound to `localhost`. Prior this was done by Theia calling `gp forward-port`
- makes Theia consume supervisor's gRPC interface to implement the `ServedPortsService`

Note: the majority of changes is in generated code

## How to test
- start a workspace
- Listen to supervisor's API:
```
curl localhost:22999/_supervisor/v1/status/ports/observe/true
```
- Start a locally bound service and observe the output of the curl and Theia reacting properly:
```
curl lama.sh | LAMA_PORT=9000 sh -s -- -l
```
- Start a globally bound service and observe the output of the curl and Theia reacting properly:
```
curl lama.sh | sh
```
